### PR TITLE
[MRG] New api design

### PR DESCRIPTION
**Pull request by wdevazelhes**
This PR merges `new_api_design` into `master`, adding the new API to metric learn as well as a minimal documentation 

Achieves #91 
Also addresses #123  

TODO: 
- [x] Change `README` ? 
<strike>- [ ] fix pep8 erros comparing (in the diff with master)</strike>
<strike>- [ ]update the notebook `metric_plotting.ipynb` with the new API</strike>
<strike>- [ ] take a second look at methods that were modified to prevent using `self.X_` (and therefore have a new argument `X`) (ex: `RCA._check_dimension`) some of them could be static methods. It could be good to explicitly make them so (maybe just file an issue for that and solve after not to mix with this PR)</strike>

### CODE REVIEW

File: test/test_transformer_metric_conversion.py [hunk 1]
```diff
@@ -19,60 +19,60 @@ def setUpClass(self):
...
     seed = np.random.RandomState(1234)
     sdml = SDML_Supervised(num_constraints=1500)
     sdml.fit(self.X, self.y, random_state=seed)
-    L = sdml.transformer()
+    L = sdml.transformer_
     assert_array_almost_equal(L.T.dot(L), sdml.metric())
 
   def test_nca(self):
     n = self.X.shape[0]
     nca = NCA(max_iter=(100000//n), learning_rate=0.01)
```
Comment (bellet) [inline test/test_transformer_metric_conversion.py]: again here

File: doc/supervised.rst [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+
+    import numpy as np
+    from metric_learn import NCA
+    from sklearn.datasets import load_iris
+
+    iris_data = load_iris()
+    X = iris_data['data']
+    Y = iris_data['target']
+
+    nca = NCA(max_iter=1000, learning_rate=0.01)
```
[the remark this answers is not in the document]
Comment (wdevazelhes) [reply doc/supervised.rst]: Indeed, thanks !
Done

File: test/test_transformer_metric_conversion.py (hunk 1 above)
Comment (wdevazelhes) [reply test/test_transformer_metric_conversion.py]: Yep, done

**Comment by wdevazelhes** [comment]
I'll also update the `README.rst` here (for instance removing the paragraph about `transformer()`)

**Comment by bellet** [comment]
Should we update the README, or do these instructions refer to the last release (install feom PyPi)?

**Comment by wdevazelhes** [comment]
> Should we update the README, or do these instructions refer to the last release (install from PyPi)?

Ah yes that's right, they seem to refer to the last release
Should we keep instructions on using the package (algorithms etc) in the `README` though ? Because this seems to be already documented in the documentation (by the way in the doc we say the distance is ``(x−y)M(x−y)`` whereas in the README we say it's ``sqrt((x−y)M(x−y))``. I guess the real distance is with sqrt so we should modify in the docs right ? )
Because if we make changes in algorithms we'll have to always update this page, but do it just after the release, it seems a bit weird

**Comment by wdevazelhes** [comment]
I just added a TODO in the description of the PR

**Comment by bellet** [comment]
I agree that we could probably remove the detailed usage instructions on the README and only keep the link to the doc
And yes, the proper distance is with the square root and this is also what is returned by `score_pairs` so we should change the doc accordingly

**Comment by perimosocordiae** [comment]
I'd prefer to keep this PR limited as much as possible to the merge, and then fix the remaining minor issues (pep8 stuff, jupyter notebook, static methods, etc) separately. The idea is that those will be small and easy to review, whereas this PR is pretty large already.

**Comment by bellet** [comment]
Yes that's true, reviewing additional things here will be more difficult
@wdevazelhes could you perhaps simply update the README and move the rest to new issues that you can deal with separately? So we can merge this quickly and start making progress on small things and on improving the doc in separate PRs

**Comment by wdevazelhes** [comment]
I agree, I just pushed the new README and updated the introduction with the square root

**Comment by wdevazelhes** [comment]
> About `self.X_` / `self.X`: if you don't do it here, make sure you remove #134 from the list of issues addressed by this PR

I replaced `self.X_` by `self.X` here, but there still remains some variables that maybe do not need to be stored after fit, like for instance `self.labels_` in ``LMNN``, or `self.vab_` in `LSML`, so I'll remove it from listed issues, and change the title to a more general title (not just "avoid storing X" but "avoid storing unnecessary variables")

**Comment by bellet** [comment]
Thanks, I will merge this
